Slash Commands

binaryoverload opened the pull request.

binaryoverload committed:
Add command meta classes and enable loading and registering of commands

weeryan17 commented on src/main/kotlin/org/cascadebot/bot/cmd/DashboardCommand.kt:
```diff
@@ -0,0 +1,18 @@
+
+    override fun onCommand(context: CommandContext, args: CommandArgs, data: Any) {
+//        context.reply("HI")
+        context.replyE("Test")
```
**E**

weeryan17 replied:
replyE method doesn't even exist lol

weeryan17 commented on src/main/kotlin/org/cascadebot/bot/events/ReadyListener.kt:
```diff
@@ -23,5 +24,7 @@ class ReadyListener(): ListenerAdapter() {
             rabbitMQManager.channel.queueBind("shard-$shardId", "amq.topic", "shard.$shardId.*.#")
         }
+
+        Main.commandManager.registerCommandsOnce(event.jda)
```
We should probably talk about this, considering we're barely going to have any global commands I suppose it's fine, just feels wrong.

weeryan17 commented on src/main/kotlin/org/cascadebot/bot/events/InteractionListener.kt:
```diff
@@ -0,0 +1,27 @@
+                args,
+                ""
+            )
+        }
```
we should probably throw an exception in the else of this if. Theoretically it should never happen, but if it does we should know about it.

weeryan17 commented on src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt:
```diff
@@ -0,0 +1,71 @@
+    val commands
+        get() = _commands.toList()
+
+    private val registeredCommands: AtomicBoolean = AtomicBoolean(false)
```
This never gets updated

binaryoverload committed:
CommandManager: Update

Set path to be string instead of POJO
Set the atomic boolean so it actually only runs once
Add extra debug log

binaryoverload committed:
Add logback util

binaryoverload committed:
Add config and CLI option for updating bot commands to Discord

binaryoverload committed:
Add config option to enable debug loggin

binaryoverload committed:
Update example config

binaryoverload committed:
Add context methods, add message types + colors, commands now work

weeryan17 commented on src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt:
```diff
@@ -0,0 +1,71 @@
+            for ((path, command) in _commands.mapKeys { it.key.path.joinToString("/") }) {
+                logger.debug("Attempting to register command '$path'")
+                try {
+                    val cmd = jda.upsertCommand(command.commandData).complete()
```
Ok I know we're not going to upload many commands, but I fell like complete is wrong to do. We should probably have a quick conversation about this

binaryoverload replied:
I've changed it to use queue, thoughts?

weeryan17 replied:
looks good, neat that you can combine rest actions like that.

weeryan17 commented on src/main/kotlin/org/cascadebot/bot/cmd/meta/CommandManager.kt:
```diff
@@ -0,0 +1,71 @@
+        // Already registered commands on another shard
+        if (registeredCommands.getAndSet(true)) return logger.debug("Commands have already been registered on another shard")
+
+        var successful = 0
```
Why does this variable exist? We (currently) don't use it anywhere

binaryoverload replied:
We do! See lines 63 and 69

weeryan17 replied:
Oh I see you use it for the log, missed that my bad.

binaryoverload committed:
Register commands using queue not complete
